File: java/toolgood.words/src/main/java/toolgood/words/internals/IntDictionary.java
package toolgood.words.internals;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class IntDictionary {
    private int[] _keys;
    private int[] _values;
    private int last;

    public IntDictionary() {
        last = -1;
    }

    public int[] getKeys() {
        return _keys;
    }

    public int[] getValues() {
        return _values;
    }

    public void SetDictionary(Map<Integer, Integer> dict) {

        List<Integer> keys = new ArrayList<Integer>();
        dict.forEach((k, v) -> {
            keys.add((int) k);
        });

        _keys = new int[dict.size()];
        _values = new int[dict.size()];
        for (int i = 0; i < keys.size(); i++) {
            _keys[i] = keys.get(i);
            _values[i] = dict.get(_keys[i]);
        }
        last = _keys.length - 1;
    }

    public void SetDictionary(int[] keys, int[] values) {
        _keys = keys;
        _values = values;
        last = _keys.length - 1;
    }

    public int IndexOf(int key) {
        if (last == -1) {
            return -1;
        }
        if (_keys[0] == key) {
            return 0;
        }
        if (_keys[last] == key) {
            return last;
        }

        int left = 0;
        int right = last;
        while (left + 1 < right) {
            int mid = (left + right) >> 1;
            int d = _keys[mid] - key;

            if (d == 0) {
                return mid;
            } else if (d > 0) {
                right = mid;
            } else {
                left = mid;
            }
        }
        return -1;
    }

    public int GetValue(int index){
        return _values[index];
    }
 
}

File: csharp/ToolGood.Words/internals/IntDictionary.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ToolGood.Words.internals
{
    public struct IntDictionary
    {
        private ushort[] _keys;
        private int[] _values;
        private int last;
        public IntDictionary(ushort[] keys, int[] values)
        {
            _keys = keys;
            _values = values;
            last = keys.Length - 1;
        }
        public IntDictionary(Dictionary<ushort, int> dict)
        {
            var keys = dict.Select(q => q.Key).OrderBy(q => q).ToArray();
            var values = new int[keys.Length];
            for (int i = 0; i < keys.Length; i++) {
                values[i] = dict[keys[i]];
            }
            _keys = keys;
            _values = values;
            last = keys.Length - 1;
        }


        public ushort[] Keys {
            get {
                return _keys;
            }
        }

        public int[] Values {
            get {
                return _values;
            }
        }

        public bool TryGetValue(ushort key, out int value)
        {
            if (last == -1) {
                value = 0;
                return false;
            }
            if (_keys[0] == key) {
                value = _values[0];
                return true;
            } else if (last == 0 || _keys[0] > key) {
                value = 0;
                return false;
            }

            if (_keys[last] == key) {
                value = _values[last];
                return true;
            } else if (_keys[last] < key) {
                value = 0;
                return false;
            }

            var left = 1;
            var right = last - 1;
            while (left <= right) {
                int mid = (left + right) >> 1;
                int d = _keys[mid] - key;

                if (d == 0) {
                    value = _values[mid];
                    return true;
                } else if (d > 0) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            value = 0;
            return false;
        }



    }
}
